pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java

```diff
@@ -0,0 +1,168 @@
+   * @throws InterruptedException if the current thread is interrupted
+   */
+  public boolean tryAcquire(long timeout, TimeUnit unit)
+      throws InterruptedException {
+    if (_maxConcurrentQueries <= 0) {
+      return true;
+    }
+    return _semaphore.tryAcquire(timeout, unit);
+  }
+
+  /**
+   * Should be called after the query is done executing. It is the responsibility of the caller to ensure that this
+   * method is called exactly once for each call to {@link #tryAcquire(long, TimeUnit)}.
+   */
+  public void release() {
+    if (_maxConcurrentQueries > 0) {
+      _semaphore.release();
+    }
+  }
+
+  @Override
+  public void processClusterChange(HelixConstants.ChangeType changeType) {
+    Preconditions.checkArgument(
+        changeType == HelixConstants.ChangeType.EXTERNAL_VIEW || changeType == HelixConstants.ChangeType.CLUSTER_CONFIG,
+        "MultiStageQuerySemaphore can only handle EXTERNAL_VIEW and CLUSTER_CONFIG changes");
+
+    if (changeType == HelixConstants.ChangeType.EXTERNAL_VIEW) {
+      List<String> clusterInstances = _helixAdmin.getInstancesInCluster(_helixManager.getClusterName());
+      int numBrokers = Math.max(1, (int) clusterInstances.stream()
+          .filter(instance -> instance.startsWith(CommonConstants.Helix.PREFIX_OF_BROKER_INSTANCE))
+          .count());
+      int numServers = Math.max(1, (int) clusterInstances.stream()
+          .filter(instance -> instance.startsWith(CommonConstants.Helix.PREFIX_OF_SERVER_INSTANCE))
+          .count());
+
+      if (numBrokers != _numBrokers || numServers != _numServers) {
+        _numBrokers = numBrokers;
+        _numServers = numServers;
+        if (_maxConcurrentQueries > 0) {
+          _semaphore.setPermits(Math.max(1, _maxConcurrentQueries * _numServers / _numBrokers));
+        }
+      }
+    } else {
+      int maxConcurrentQueries = Integer.parseInt(
+          _helixAdmin.getConfig(_helixConfigScope,
+                  Collections.singletonList(CommonConstants.Helix.CONFIG_OF_MAX_CONCURRENT_MULTI_STAGE_QUERIES))
+              .getOrDefault(CommonConstants.Helix.CONFIG_OF_MAX_CONCURRENT_MULTI_STAGE_QUERIES,
+                  CommonConstants.Helix.DEFAULT_MAX_CONCURRENT_MULTI_STAGE_QUERIES));
+
+      if (_maxConcurrentQueries == maxConcurrentQueries) {
+        return;
+      }
+
+      if (_maxConcurrentQueries <= 0 && maxConcurrentQueries > 0
+          || _maxConcurrentQueries > 0 && maxConcurrentQueries <= 0) {
+        // This operation isn't safe to do while queries are running so we require a restart of the broker for this
+        // change to take effect.
+        LOGGER.warn("Enabling or disabling limitation of the maximum number of multi-stage queries running "
+            + "concurrently requires a restart of the broker to take effect");
+        return;
+      }
+
+      if (maxConcurrentQueries > 0) {
+        _semaphore.setPermits(Math.max(1, maxConcurrentQueries * _numServers / _numBrokers));
+      }
```

Comment:
I think there is a possible NPE here. Imagine that the broker is initialized with -1 concurrent queries, `_semaphore` is kept as null. Then the config is changed to set an actual max concurrent queries, then we try to call `_semaphore.setPermits` here without actually creating the semaphore.

For simplicity, I would suggest to always allocate the semaphore (with whatever permits). Given don't delegate on the semaphore when `_maxConcurrentQueries <= 0`, that should be a clean and correct solution

Reply:
Maybe we can add a test to verify my scenario.

Reply:
> I think there is a possible NPE here. Imagine that the broker is initialized with -1 concurrent queries, _semaphore is kept as null. Then the config is changed to set an actual max concurrent queries, then we try to call _semaphore.setPermits here without actually creating the semaphore.

Unless I'm misunderstanding you, that condition is already being checked just before this line and we're disallowing such a change right? This makes things a lot more simple to reason about - transitioning from disabled to enabled or vice versa while queries are running will create lots of edge cases to handle with the acquire / release mechanism.

> Maybe we can add a test to verify my scenario.

There's already tests verifying disabled -> enabled and enabled -> disabled transitions.

Reply:
Oh! I didn't _parse_ the previous check. It makes sense. In that case queries that were running when the feature was disabled will suddenly try to release the permit once they finish.

I guess we can resolve that by keeping a reference to a virtual _semaphore_ (not an actual Semaphore, but something that implements the API) and then call release on that object, which will be a noop on queries started when the feature was disabled.

But it is true that it is not a blocking issue and we can assume it requires a restart in this case.

But... isn't that actually going to be problematic when moving from N to M permits? I mean:

1. The semaphore is configured with N (where N > 0) permits.
1. Query starts, semaphore has N (where N > 0) permits, so it consumes one and continues. The semaphore has now N-1 permits.
2. The config changes and now we need to use use M permits. We restart the semaphore to use that number.
3. The query ends and release its permit. Now the semaphore has M+1 permits.

In the worst case scenario we could end up having N+M permits. Luckily, this is not accumulative. But unless I'm wrong I think we need to fix the issue.

Reply:
I have two proposals to fix it:
1. Modify `tryAcquire` to return an object that:
  * has a method to check whether we are allowed to go
  * is AutoCloseable. Its close method release the lock if and only if the semaphore didn't change (we can use a version stored in a AtomicInt to verify that without locking).
2. Keep in this object a set of threads and modify the release method to only release the permit if the caller thread is listed in the set. We can clean the set when config changes. This requires less changes in the code, but the synchronization may be expensive.

In general I think we need to synchronize in this class anyway

Reply:
> But... isn't that actually going to be problematic when moving from N to M permits

Hmm, I don't think so - we're increasing / decreasing the number of permits based on the previous total number of permits, not the current available number of permits. You can see some of the unit tests like `testIncreaseNumBrokers` to see how this would look like in action.

> In general I think we need to synchronize in this class anyway

The cluster change handling should be done in a single threaded fashion and the semaphore should support concurrent modification of the number of permits while other threads are trying to acquire / release so I don't think additional external synchronization is necessary here?

Reply:
You are right. I though we were calling a _set_ method to change the number of permits instead of sending the increase/decrease.